README.md
<p align="center">
<img src="src/dashboard_pages/imgs/logo.png" alt="HAR-Tastic" width="150"/>
</p>

# HAR-TASTIC

HAR-Tastic is Human Activity Recognition (HAR) model that can detect what activity you are performing based on your phone's accelerometers' readings. 

It has been developed as final project for the Machine Learnign and Big Data Bootcamp @ [CORE School](https://www.corecode.school/).

### Access the Project [here](https://share.streamlit.io/aesqgr/har-fbc/main/dashboard.py)

## About the project

Have you ever wondered how apps like Runtastic or Strava are able to detect what kind of activity you are performing? Whether is running, sleeping or simply walking, your phone or wearable are able to detect what you are doing.

I have implemented a Long Short-term memory, an artificial recurrent neural network architecture for the human activity recognition task.

## About the data 

Your phone is packed with sensors that can be used to monitor your activity. Specifically, they contain tri-axial accelerometers that measure acceleration in all three spatial dimensions. 

The Dataset is obtained from WISDM Lab, Department of Computer & Information Science, Fordham University, NY [here]("https://www.cis.fordham.edu/wisdm/dataset.php"). It was collected from 36 uses as they performed different activities. 


## About the model

From Wikipedia:
``` 
Long short-term memory (LSTM) is an artificial recurrent neural network (RNN) architecture \n
used in the field of deep learning. Unlike standard feedforward neural networks, LSTM has \n
feedback connections. It can process not only single data points (such as images), but also \n
entire sequences of data (such as speech or video). For example, LSTM is applicable to tasks \n
such as unsegmented, connected handwriting recognition, speech recognition and anomaly detection \n
in network traffic or IDSs (intrusion detection systems).
```

src/history
,loss,accuracy,val_loss,val_accuracy
0,0.0658838301897049,0.9782893657684326,0.11854473501443863,0.9641808271408081
1,0.06363965570926666,0.979469895362854,0.12165538966655731,0.964411199092865
2,0.05921642482280731,0.9808376431465149,0.13683658838272095,0.9609559178352356
3,0.05965518206357956,0.9807800054550171,0.12924160063266754,0.9630290865898132
4,0.05954853445291519,0.9802905321121216,0.12816445529460907,0.9641232490539551
5,0.06006208062171936,0.9808232188224792,0.12273502349853516,0.9655629396438599
6,0.05848778039216995,0.9809671640396118,0.12570200860500336,0.9643535614013672
7,0.05532286316156387,0.9818885922431946,0.13153879344463348,0.9629715085029602
8,0.05432158708572388,0.9824500679969788,0.1262640506029129,0.9638928771018982
9,0.051096878945827484,0.9837026000022888,0.1320422887802124,0.9652174115180969
10,0.0516546405851841,0.9832707047462463,0.12830087542533875,0.9651021957397461
11,0.049633074551820755,0.9839473366737366,0.12210605293512344,0.965793251991272
12,0.046882253140211105,0.9851422905921936,0.1336773782968521,0.9642384052276611
13,0.046208932995796204,0.9847391843795776,0.13711333274841309,0.9634321928024292
14,0.046173468232154846,0.9848255515098572,0.12418010085821152,0.9666570425033569
15,0.044347915798425674,0.9857038259506226,0.12013272196054459,0.9663691520690918
16,0.04466890171170235,0.9853294491767883,0.13446520268917084,0.9651021957397461
17,0.043784767389297485,0.9859485626220703,0.14042867720127106,0.9642959833145142
18,0.04106825962662697,0.9868555665016174,0.13395026326179504,0.9666570425033569
19,0.03849584981799126,0.9872874617576599,0.14019912481307983,0.96498703956604
20,0.04011497646570206,0.9870283007621765,0.140060693025589,0.9655053019523621
21,0.03833157569169998,0.9874602556228638,0.13613028824329376,0.9676360487937927
22,0.03616449981927872,0.9886119961738586,0.14202997088432312,0.9659660458564758
23,0.037593696266412735,0.9881080985069275,0.13279366493225098,0.9671753644943237
24,0.03648122772574425,0.9881513118743896,0.15050992369651794,0.9645839333534241
25,0.03572091460227966,0.9883672595024109,0.13187655806541443,0.9677512049674988
26,0.032831065356731415,0.9894038438796997,0.14042851328849792,0.9651021957397461
27,0.03218808397650719,0.9893750548362732,0.1434413343667984,0.965447723865509
28,0.04272780939936638,0.9864380359649658,0.12906885147094727,0.9676936268806458
29,0.030355552211403847,0.9904835820198059,0.13928498327732086,0.9661963582038879
30,0.034044332802295685,0.9888999462127686,0.13648173213005066,0.9670025706291199
31,0.02923300862312317,0.9904547929763794,0.13442522287368774,0.9676360487937927
32,0.029978260397911072,0.9902676343917847,0.15142033994197845,0.9665994644165039
33,0.035078052431344986,0.9888999462127686,0.15655522048473358,0.9631442427635193
34,0.028641048818826675,0.9909011125564575,0.13641998171806335,0.9684422612190247
35,0.027853038161993027,0.9909154772758484,0.1472049206495285,0.9669449925422668
36,0.029176022857427597,0.9909443259239197,0.13489897549152374,0.9701123237609863
37,0.025537772104144096,0.9916209578514099,0.14724254608154297,0.9664267301559448
38,0.022779757156968117,0.9929167032241821,0.15763217210769653,0.9675784707069397
39,0.024409757927060127,0.9921248555183411,0.15963608026504517,0.9673480987548828
40,0.025838298723101616,0.9914625883102417,0.1763199269771576,0.9646990895271301
41,0.028651032596826553,0.9906563758850098,0.15625332295894623,0.9655053019523621
42,0.029616765677928925,0.9906563758850098,0.14846627414226532,0.964871883392334
43,0.023638511076569557,0.9922544360160828,0.15375296771526337,0.9680967330932617
44,0.021652519702911377,0.9929454922676086,0.1453203707933426,0.9687302112579346
45,0.022831732407212257,0.9923695921897888,0.16755644977092743,0.9633746147155762
46,0.02397705614566803,0.9919520616531372,0.15450187027454376,0.9679815769195557
47,0.024211065843701363,0.9922975897789001,0.15280713140964508,0.9667722582817078
48,0.024373507127165794,0.9920240640640259,0.1474296897649765,0.9690757393836975
49,0.022286048159003258,0.9930318593978882,0.15410012006759644,0.965850830078125


src/dashboard_pages/test.py
import streamlit as st
import json
import json
import pandas as pd
import sys
from matplotlib import pyplot as plt

from streamlit_folium import folium_static
def app():
    st.title('Population')
    st.write('This is the *population* page.')
    columns_select = st.beta_columns(2)
    with columns_select[0]:
        district_option = st.selectbox('Select Disctrict', "test")
   

src/model_fld/model_build.py
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import scipy
from sklearn.model_selection import train_test_split
from keras.models import Sequential, load_model
from keras.layers import LSTM, Dense, Flatten, Dropout
warnings.filterwarnings('ignore')
from datacleaning import reshaped_segments, labels, random_seed,n_epochs,batch_size

#split test
X_train, X_test, y_train, y_test = train_test_split(reshaped_segments, labels, test_size = 0.2, random_state = random_seed)

#model architecture
model = Sequential()
# RNN layer
model.add(LSTM(units = 128, input_shape = (X_train.shape[1], X_train.shape[2])))
# Dropout layer
model.add(Dropout(0.5)) 
# Dense layer with ReLu
model.add(Dense(units = 64, activation='relu'))
# Softmax layer
model.add(Dense(y_train.shape[1], activation = 'softmax'))
# Compile model
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

#train model
history = model.fit(X_train, y_train, epochs = n_epochs, validation_split = 0.20, batch_size = batch_size, verbose = 1)

#save model and model history
history.to_csv(index=False)
model.save("model.h5")

src/model_fld/datacleaning.py
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import scipy
from sklearn.model_selection import train_test_split
from keras.models import Sequential, load_model
from keras.layers import LSTM, Dense, Flatten, Dropout
warnings.filterwarnings('ignore')
import sys
sys.path.append("..")

#build dataframe
columns = ['user','activity','timestamp', 'x-axis', 'y-axis', 'z-axis']
df_har = pd.read_csv('../WISDM_ar_v1.1/WISDM_ar_v1.1_raw.txt', header = None, names = columns)

#clean data
df_har = df_har.dropna()
df_har.shape
df_har['z-axis'] = df_har['z-axis'].str.replace(';', '')
df_har['z-axis'] = df_har['z-axis'].apply(lambda x:float(x))
df = df_har[df_har['timestamp'] != 0]
df = df.sort_values(by = ['user', 'timestamp'], ignore_index=True)


# build segments
random_seed = 42   
n_time_steps = 50 
n_features = 3 
step = 10 
n_classes = 6 
n_epochs = 50       
batch_size = 1024   
learning_rate = 0.0025
l2_loss = 0.0015
segments = []
labels = []

for i in range(0,  df.shape[0]- n_time_steps, step):  

    xs = df['x-axis'].values[i: i + 50]

    ys = df['y-axis'].values[i: i + 50]

    zs = df['z-axis'].values[i: i + 50]

    label = scipy.stats.mode(df['activity'][i: i + 50])[0][0]

    segments.append([xs, ys, zs])

    labels.append(label)

#reshape segments
reshaped_segments = np.asarray(segments, dtype= np.float32).reshape(-1, n_time_steps, n_features)
labels = np.asarray(pd.get_dummies(labels), dtype = np.float32)

def segm():
    return reshaped_segments

src/dashboard_pages/model_dash.py
import sys
print(sys.path)
sys.path.append("..")
sys.path.append("../src")
import matplotlib.pyplot as plt
import pandas as pd
import streamlit as st
import seaborn as sns
from keras.models import load_model, Sequential
from keras.layers import LSTM, Dense, Flatten, Dropout
import numpy as np
import tensorflow as tf
from sklearn.model_selection import train_test_split



model = load_model('src/model_fld/model.h5')


#build dataframe
columns = ['user','activity','timestamp', 'x-axis', 'y-axis', 'z-axis']
df_har = pd.read_csv('./src/WISDM_ar_v1.1/WISDM_ar_v1.1_raw.txt', header = None, names = columns)
df_har = df_har.dropna()
df_har.shape
df_har['z-axis'] = df_har['z-axis'].str.replace(';', '')
df_har['z-axis'] = df_har['z-axis'].apply(lambda x:float(x))
df = df_har[df_har['timestamp'] != 0]
df = df.sort_values(by = ['user', 'timestamp'], ignore_index=True)

#misc
history = pd.read_csv('./src/model_fld/history_test.csv', header = [0])


def app():
    st.set_option('deprecation.showPyplotGlobalUse', False)
    st.title('Data Cleaning & Model architecture')
    st.write('Echemos un vistazo a los datos...')
    st.write("**Original shape:** ", df_har.shape)
    st.write("**Final shape:** ", df.shape)
    st.write("**DataFrame reduction:** ", "{:.0%}".format((df.shape[0]/df_har.shape[0] - 1)))
    st.write(df.head())

    st.write("Veamos como se reparten las muestras por cada actividad")
    sns.set_style("whitegrid")
    st.write(sns.countplot(x = "activity", data = df))
    plt.title("Number of samples by activity")
    st.pyplot(plt.show())

    st.write("Cada actividad tiene una forma característica")
    data36 = df[(df["user"] == 36) & (df["activity"] == "Jogging")][:400]
    columns_select = st.beta_columns(2)
    with columns_select[0]:
        data36 = df[(df["user"] == 36) & (df["activity"] == "Jogging")][:400]
        sns.lineplot(y = "x-axis", x = "timestamp", data = data36)
        sns.lineplot(y = "y-axis", x = "timestamp", data = data36)
        sns.lineplot(y = "z-axis", x = "timestamp", data = data36)
        plt.legend(["x-axis", "y-axis", "z-axis"])
        plt.ylabel("Jogging")
        plt.title("Jogging", fontsize = 15)
        st.pyplot()
        data36 = df[(df["user"] == 36) & (df["activity"] == "Walking")][:400]
        sns.lineplot(y = "x-axis", x = "timestamp", data = data36)
        sns.lineplot(y = "y-axis", x = "timestamp", data = data36)
        sns.lineplot(y = "z-axis", x = "timestamp", data = data36)
        plt.legend(["x-axis", "y-axis", "z-axis"])
        plt.ylabel("Walking")
        plt.title("Walking", fontsize = 15)
        st.pyplot()
        data36 = df[(df["user"] == 36) & (df["activity"] == "Upstairs")][:400]
        sns.lineplot(y = "x-axis", x = "timestamp", data = data36)
        sns.lineplot(y = "y-axis", x = "timestamp", data = data36)
        sns.lineplot(y = "z-axis", x = "timestamp", data = data36)
        plt.legend(["x-axis", "y-axis", "z-axis"])
        plt.ylabel("Upstairs")
        plt.title("Upstairs", fontsize = 15)
        st.pyplot()
    with columns_select[1]:
        data36 = df[(df["user"] == 36) & (df["activity"] == "Downstairs")][:400]
        sns.lineplot(y = "x-axis", x = "timestamp", data = data36)
        sns.lineplot(y = "y-axis", x = "timestamp", data = data36)
        sns.lineplot(y = "z-axis", x = "timestamp", data = data36)
        plt.legend(["x-axis", "y-axis", "z-axis"])
        plt.ylabel("Downstairs")
        plt.title("Downstairs", fontsize = 15)
        st.pyplot()
        data36 = df[(df["user"] == 36) & (df["activity"] == "Standing")][:400]
        sns.lineplot(y = "x-axis", x = "timestamp", data = data36)
        sns.lineplot(y = "y-axis", x = "timestamp", data = data36)
        sns.lineplot(y = "z-axis", x = "timestamp", data = data36)
        plt.legend(["x-axis", "y-axis", "z-axis"])
        plt.ylabel("Standing")
        plt.title("Standing", fontsize = 15)
        st.pyplot()
        data36 = df[(df["user"] == 36) & (df["activity"] == "Sitting")][:400]
        sns.lineplot(y = "x-axis", x = "timestamp", data = data36)
        sns.lineplot(y = "y-axis", x = "timestamp", data = data36)
        sns.lineplot(y = "z-axis", x = "timestamp", data = data36)
        plt.legend(["x-axis", "y-axis", "z-axis"])
        plt.ylabel("Sitting")
        plt.title("Sitting", fontsize = 15)
        st.pyplot()
        
    st.image('src/dashboard_pages/imgs/model_summary.png')

    plt.plot(np.array(history['loss']), "r--", label = "Train loss")
    plt.plot(np.array(history['accuracy']), "g--", label = "Train accuracy")
    plt.plot(np.array(history['val_loss']), "r-", label = "Validation loss")
    plt.plot(np.array(history['val_accuracy']), "g-", label = "Validation accuracy")
    plt.title("Training session's progress over iterations")
    plt.legend(loc='lower left')
    plt.ylabel('Training Progress (Loss/Accuracy)')
    plt.xlabel('Training Epoch')
    plt.ylim(0) 
    st.pyplot()
    
    st.image('src/dashboard_pages/imgs/conf_matrix.png')


    